src/java.base/share/classes/java/lang/reflect/ProxyGenerator.java

```diff
@@ -597,6 +598,13 @@ private void generateStaticInitializer() {
         mv.visitTryCatchBlock(L_startBlock, L_endBlock, L_NoClassHandler,
                 JL_CLASS_NOT_FOUND_EX);
 
+        // Put ClassLoader at local variable index 0, used by
+        // Class.forName(String, boolean, ClassLoader) calls
+        mv.visitLdcInsn(Type.getObjectType(dotToSlash(className)));
+        mv.visitMethodInsn(INVOKEVIRTUAL, JL_CLASS,
+                "getClassLoader", "()" + LJL_CLASSLOADER, false);
```

Comment:
Hello @liach, should this instead be using the (application supplied) `loader` returned by the call to `ProxyGenerator.getClassLoader()` or maybe the `loader` member in the `ProxyGenerator` itself?

Reply:
This is equivalent to `jdk.proxy5.$Proxy5.class.getClassLoader()` in Java source code, so this is exactly the application-supplied loader, which also uses the same loader as the previous behavior of `forName` calls.

If you want to pass the loader from `ProxyGenerator` to the proxy, it requires complex tricks. Hidden classes won't work due to serialization incompatibility; accessor methods would be defined in `jdk.internal` and exported specifically to the proxy modules, but writing a class so each proxy gets its loader while what I wrote can already do is overkill.

Reply:
Thank you @liach for that detail.